Fix uninitialized string drop on panic in let_cxx_string expression.

Fixes unsoundness reported in https://github.com/dtolnay/cxx/issues/1729
and resolved in https://github.com/dtolnay/cxx/pull/1731 (cxx version 1.0.195).

Note: Once cxx is upgraded past 1.0.194 (to 1.0.195 or later), this patch can be dropped.

--- a/src/cxx_string.rs
+++ b/src/cxx_string.rs
@@ -4,6 +4,7 @@ use crate::lossy;
 use alloc::borrow::Cow;
 #[cfg(feature = "alloc")]
 use alloc::string::String;
+use core::cell::UnsafeCell;
 use core::cmp::Ordering;
 use core::ffi::{c_char, CStr};
 use core::fmt::{self, Debug, Display};
@@ -84,11 +85,13 @@ pub struct CxxString {
 #[macro_export]
 macro_rules! let_cxx_string {
     ($var:ident = $value:expr $(,)?) => {
-        let mut cxx_stack_string = $crate::private::StackString::new();
+        let cxx_stack_string = $crate::private::StackString::new();
         #[allow(unused_mut, unused_unsafe)]
         let mut $var = match $value {
             let_cxx_string => unsafe { cxx_stack_string.init(let_cxx_string) },
         };
+        #[allow(unused_unsafe)]
+        let _cxx_stack_string_drop_guard = unsafe { cxx_stack_string.drop_guard() };
     };
 }
 
@@ -301,32 +304,41 @@ impl std::io::Write for Pin<&mut CxxString> {
 pub struct StackString {
     // Static assertions in cxx.cc validate that this is large enough and
     // aligned enough.
-    space: MaybeUninit<[usize; 8]>,
+    space: UnsafeCell<MaybeUninit<[usize; 8]>>,
 }
 
+unsafe impl Sync for StackString {}
+
 impl StackString {
     pub fn new() -> Self {
         StackString {
-            space: MaybeUninit::uninit(),
+            space: UnsafeCell::new(MaybeUninit::uninit()),
         }
     }
 
-    pub unsafe fn init(&mut self, value: impl AsRef<[u8]>) -> Pin<&mut CxxString> {
+    #[allow(clippy::mut_from_ref)]
+    pub unsafe fn init(&self, value: impl AsRef<[u8]>) -> Pin<&mut CxxString> {
         let value = value.as_ref();
         unsafe {
-            let this = &mut *self.space.as_mut_ptr().cast::<MaybeUninit<CxxString>>();
+            let this = &mut *self.space.get().cast::<MaybeUninit<CxxString>>();
             string_init(this, value.as_ptr(), value.len());
             Pin::new_unchecked(&mut *this.as_mut_ptr())
         }
     }
-}
 
-impl Drop for StackString {
-    fn drop(&mut self) {
-        unsafe {
-            let this = &mut *self.space.as_mut_ptr().cast::<MaybeUninit<CxxString>>();
-            string_destroy(this);
+    pub unsafe fn drop_guard(&self) -> impl Drop + '_ {
+        struct StackStringDropGuard<'a>(&'a StackString);
+
+        impl<'a> Drop for StackStringDropGuard<'a> {
+            fn drop(&mut self) {
+                unsafe {
+                    let this = &mut *self.0.space.get().cast::<MaybeUninit<CxxString>>();
+                    string_destroy(this);
+                }
+            }
         }
+
+        StackStringDropGuard(self)
     }
 }
